velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp

```diff
@@ -0,0 +1,353 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.h"
+#include "velox/common/file/File.h"
+#include "velox/connectors/hive/HiveConfig.h"
+#include "velox/connectors/hive/storage_adapters/gcs/GCSUtil.h"
+#include "velox/core/Config.h"
+
+#include <fmt/format.h>
+#include <glog/logging.h>
+#include <memory>
+#include <stdexcept>
+
+#include <google/cloud/storage/client.h>
+
+namespace facebook::velox {
+namespace {
+namespace gcs = ::google::cloud::storage;
+namespace gc = ::google::cloud;
+// Reference: https://issues.apache.org/jira/browse/ARROW-14346
+// https://github.com/apache/arrow/blob/master/cpp/src/arrow/filesystem/gcsfs.cc#L49
+// Change the default upload buffer size. In general, sending larger buffers is
+// more efficient with GCS, as each buffer requires a roundtrip to the service.
+// With formatted output (when using `operator<<`), keeping a larger buffer in
+// memory before uploading makes sense.  With unformatted output (the only
+// choice given gcs::io::OutputStream's API) it is better to let the caller
+// provide as large a buffer as they want. The GCS C++ client library will
+// upload this buffer with zero copies if possible.
+auto constexpr kUploadBufferSize = 256 * 1024;
+
+class GCSReadFile final : public ReadFile {
+ public:
+  GCSReadFile(const std::string& path, gcs::Client* client) : client_(client) {
+    // assumption it's a proper path
+    bucketAndKeyFromGCSPath(path, bucket_, key_);
+  }
+
+  // Gets the length of the file.
+  // Checks if there are any issues reading the file.
+  void initialize() {
+    // Make it a no-op if invoked twice.
+    if (length_ != -1) {
+      return;
+    }
+    // get metadata and initialize length
+    auto metadata = client_->GetObjectMetadata(bucket_, key_);
+    VELOX_CHECK_GCS_OUTCOME(
```

Comment:
I will research and document with my findinds and add guards if needed.

Reply:
It may be null, but the macro VELOX_CHECK_GCS_OUTCOME checks it and in turn invokes VELOX_FAIL. However, it is difficult to follow so I will remove its usage.